### PR TITLE
Feature 57 comments and reactions

### PR DESCRIPTION
**Pull request by shmato4ek**

### CODE REVIEW

File: backend/Recodo.BLL/Services/CommentService.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                .Include(comment => comment.Author)
+                    .ThenInclude(user => user.AvatarLink)
+                .FirstAsync(comment => comment.Id == commentEntity.Id);
+
+            return _mapper.Map<CommentDTO>(createdComment);
+        }
+
+        public async Task DeleteComment(int commentId)
+        {
+            var commentEntity = await _context.Comments.FirstOrDefaultAsync(p => p.Id == commentId);
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/CommentService.cs:36]: It's better to check owner (user) everywhere, as theoretically any user can change any comment by id via postmen

File: backend/Recodo.BLL/Services/ReactionsService.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        public ReactionService(RecodoDbContext context, IMapper mapper) : base(context, mapper) 
+        { }
+
+        public async Task ReactVideo(NewVideoReactionDTO reaction)
+        {
+            var video = await _context.Videos.FindAsync(reaction.VideoId);
+            if (video.Reactions.Any())
+            {
+                if (reaction.Reaction == video.Reactions.FirstOrDefault().Reaction)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/ReactionsService.cs]: If more than 1 reaction is presented?

File: backend/Recodo.BLL/Services/ReactionsService.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                UserId = reaction.UserId
+            };
+            video.Reactions.Add(newReaction);
+            await _context.SaveChangesAsync();
+            var createdReactionDTO = _mapper.Map<VideoReactionDTO>(newReaction);            
+        }
+        
+        public async Task ReactComment(NewCommentReactionDTO reaction)
+        {
+            var reactions = _context.CommentReactions.Where(x => x.UserId == reaction.UserId && x.CommentId == reaction.CommentId);
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/ReactionsService.cs]: Where -> FirstOrDefault

File: backend/Recodo.DAL/Migrations/20220317133458_AddVideoToComments.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace Recodo.DAL.Migrations
+{
+    public partial class AddVideoToComments : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.DAL/Migrations/20220317133458_AddVideoToComments.cs:7]: Empty migration?

File: frontend/src/app/app-routing.module.ts [hunk 5]
```diff
@@ -7,12 +7,12 @@ const routes: Routes = [
   {
     path: 'personal',
     loadChildren: () => import('./modules/workspace/workspace.module').then(m => m.WorkspaceModule),
-    canActivate : [TokenExistsGuard]
+    // canActivate : [TokenExistsGuard]
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/app-routing.module.ts]: Don't forget to get back

File: frontend/src/app/modules/workspace/video/comment-reactions/comment-reactions.component.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+          ReactionType.Dislike,
+          this.user
+        );
+        break;
+      default:
+        break;
+    }
+  }
+
+  public GetReactions(reactionNumber: number) {
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/workspace/video/comment-reactions/comment-reactions.component.ts:49]: You can get by type and assign this type on loading data

File: frontend/src/app/services/comment-reaction.service.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  public deleteReaction(currentUser: User, currentComment: Comment) {
+    const foundReaction = currentComment.reactions.find(
+      (reaction) =>
+        reaction.userId === currentUser.id &&
+        reaction.commentId === currentComment.id
+    );
+    currentComment.reactions.filter((reaction) => reaction != foundReaction);
+    return foundReaction?.id;
+  }
+}
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/services/comment-reaction.service.ts:70]: As for me, you build too difficult logic to work with reactions

**Review by Anatolii-Ivanov — COMMENTED** [review]
And this file again
frontend/src/assets/himn.mp4

File: package.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (Jyuart) [inline package.json]: Remove

File: package-lock.json [hunk 9]
```diff
@@ -0,0 +1,1076 @@
+{
```
Comment (Jyuart) [inline package-lock.json]: Remove

File: backend/Recodo.API/Controllers/VideoController.cs [hunk 10]
```diff
@@ -54,5 +65,25 @@ public async Task<ActionResult> Update([FromBody] UpdateVideoDTO videoDTO)
             await _videoService.Update(videoDTO);
             return NoContent();
         }
+
+        [HttpDelete("react")]
+        public async Task<IActionResult> DeleteReaction(int videoId, Reaction reaction)
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/VideoController.cs]: I suppose it's not used anymore

File: backend/Recodo.API/Controllers/CommentController.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        public async Task<IActionResult> UpdateComments(CommentDTO comment)
+        {
+            await _commentService.UpdateComment(comment);
+            return NoContent();
+        }
+
+        [HttpPost]
+        public async Task<ActionResult<CommentDTO>> CreateComment(NewCommentDTO comment)
+        {
+            return Ok(await _commentService.CreateComment(comment));
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/CommentController.cs]: Do you really need to return it on post?

File: backend/Recodo.BLL/Services/CommentService.cs (hunk 1 above)
Comment (Jyuart) [reply backend/Recodo.BLL/Services/CommentService.cs:36]: You still need to fix it. User shouldn't be able to delete comments of other users

File: backend/Recodo.BLL/Services/CommentService.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            var commentEntity = _mapper.Map<Comment>(newComment);
+            
+            _context.Comments.Add(commentEntity);
+            await _context.SaveChangesAsync();
+
+            var createdComment = await _context.Comments
+                .Include(comment => comment.Author)
+                .FirstAsync(comment => comment.Id == commentEntity.Id);
+
+            return _mapper.Map<CommentDTO>(createdComment);
```
Comment (Jyuart) [inline backend/Recodo.BLL/Services/CommentService.cs:31]: Is it possible to map from `NewCommentDTO` without retrieving it from the db?

File: backend/Recodo.BLL/Services/VideoService.cs [hunk 13]
```diff
@@ -96,5 +98,26 @@ public async Task Update(UpdateVideoDTO videoDTO)
...
+            var videoComments = await _commentService.GetAllVideosComments(videoDto.Id);
+            videoDto.Comments = videoComments;
+            return videoDto;
+        }
+
+        public async Task<VideoDTO> AddVideo (NewVideoDTO newVideo)
+        {
+            var video = _mapper.Map<Video>(newVideo);
+            video.CreatedAt = DateTime.Now;
+            video.Reactions = _mapper.Map<List<VideoReaction>>(newVideo.Reactions);
```
Comment (Jyuart) [inline backend/Recodo.BLL/Services/VideoService.cs:120]: Is it possible that video will be created already with some comments?

File: backend/Recodo.Common/Recodo.Common.csproj [hunk 14]
```diff
@@ -6,6 +6,10 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Cryptography.KeyDerivation" Version="6.0.2" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="5.0.13">
```
Comment (Jyuart) [inline backend/Recodo.Common/Recodo.Common.csproj]: Not sure it's needed here

File: backend/Recodo.DAL/Context/RecodoDbContext.cs [hunk 15]
```diff
@@ -19,6 +19,9 @@ public class RecodoDbContext : DbContext
         public DbSet<TeamInvitation> TeamInvitations { get; set; }
         public DbSet<Folder> Folders { get; set; }
         public DbSet<Comment> Comments { get; set; }
+        public DbSet<CommentReaction> CommentReactions { get; set; }
```
Comment (Jyuart) [inline backend/Recodo.DAL/Context/RecodoDbContext.cs:22]: No. It's configured as an owned entity of the `Comment`

File: backend/Recodo.DAL/Entities/Comment.cs [hunk 16]
```diff
@@ -13,6 +13,7 @@ public class Comment
         public int AuthorId { get; set; }
         public User Author { get; set; }
         public int VideoId { get; set; }
+        public Video Video { get; set; }
```
Comment (Jyuart) [inline backend/Recodo.DAL/Entities/Comment.cs:16]: Why do you need it?

File: frontend/src/app/modules/workspace/video/comment-reactions/comment-reactions.component.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+@Component({
+  selector: 'app-comment-reactions',
+  templateUrl: './comment-reactions.component.html',
+  styleUrls: ['./comment-reactions.component.scss'],
+})
+export class CommentReactionsComponent implements OnInit {
+  @Input() public comment: Comment;
+  @Input() public user: UserDto;
+  @Output() newReaction = new EventEmitter<boolean>();
+  public currentUser: UserDto;
```
Comment (Jyuart) [inline frontend/src/app/modules/workspace/video/comment-reactions/comment-reactions.component.ts:18]: It'd be nice to pass just the user id, not the whole object

File: frontend/src/app/modules/workspace/video/video-page/video-page.component.ts [hunk 18]
```diff
@@ -1,13 +1,117 @@
...
+    );
+  }
+
+  public isCurrentVideo() {
+    if (this.currentVideo) {
+      return true;
+    } else {
+      return false;
+    }
+  }
```
Comment (Jyuart) [inline frontend/src/app/modules/workspace/video/video-page/video-page.component.ts]:
```suggestion
```

File: frontend/src/app/modules/workspace/video/video-page/video-page.component.html [hunk 19]
```diff
@@ -1,16 +1,33 @@
...
+      Share
     </button>
     <button mat-icon-button>
-      <mat-icon svgIcon="share-link"></mat-icon>
+      <mat-icon svgIcon="details"></mat-icon>
     </button>
   </div>
   <app-video-player></app-video-player>
-  <app-video-reactions class="reactions-container"></app-video-reactions>
+  <div *ngIf="isCurrentVideo()">
```
Comment (Jyuart) [inline frontend/src/app/modules/workspace/video/video-page/video-page.component.html]:
```suggestion
  <div *ngIf="currentVideo && currentUser">
```

Comment (Jyuart) [reply frontend/src/app/modules/workspace/video/video-page/video-page.component.html]: But this won't work until you add getting of the current user in this component.
See, you have this property here and pass it to the child, but never initialize. 
Missing `currentUser` is a reason for that missing `avatarLink` error

File: frontend/src/app/modules/workspace/video/video-page/video-page.component.ts [hunk 20]
```diff
@@ -1,13 +1,117 @@
...
+
+  public updateVideo() {
+    this.videoService
+      .getVideoById(this.videoId)
+      .pipe(takeUntil(this.unsubscribe$))
+      .subscribe((resp) => {
+        if (resp.body != null) {
+          this.currentVideo = resp.body;
+          console.log(resp.body);
```
Comment (Jyuart) [inline frontend/src/app/modules/workspace/video/video-page/video-page.component.ts]: Remove

**Review by Jyuart — CHANGES_REQUESTED** [review]
We have a pretty large nesting

Video:
- Reactions
- Comments:
  - CommentReactions
  
And there is a problem (slightly visible refresh of the page) when just a single reaction was added to the comment.
To fix this we need not just update the video each time it requires updating, but react correspondingly. For example, after adding new reaction, we make a request to get the full video information, but update just that part that actually changed (like reactions, comments or comments reactions)

File: backend/Recodo.BLL/Services/CommentService.cs (hunk 12 above)
Comment (shmato4ek) [reply backend/Recodo.BLL/Services/CommentService.cs:31]: I tried, but it didn`t work without it

File: backend/Recodo.DAL/Context/RecodoDbContext.cs (hunk 15 above)
Comment (shmato4ek) [reply backend/Recodo.DAL/Context/RecodoDbContext.cs:22]: Yes, but if i delete this line, it will not work, don`t know why

**Pull request merged**